book/src/validity-predicates.md

```diff
@@ -1,122 +1,61 @@
+Validity predicates in principle can be both transparent and shielded. Transparent VPs are represented as WASM code and publicly visible, 
+shielded VPs are represented as arithmetic circuits hidden under ZKPs. Each transaction in Taiga has VP proofs attached to it, 
+and whoever has the verifying key (VK), can verify the proofs.
 
-* Field addition/multiplication
-* Elliptic curve addition and scalar multiplication
-* Poseidon hash
+![img.png](images/vp_img.png)
 
-#### Addresses
+Taiga uses the Halo2 proving system and validity predicates are represented as [PLONKish circuits](https://zcash.github.io/halo2/concepts/arithmetization.html). 
+For privacy reasons, all Taiga VPs share the same configuration (the set of building blocks available for the circuits), 
+and distinct VPs are created by specifying the *selectors* (that select which blocks to use).
 
-Even though Taiga does not have *accounts*, it does have *addresses*. Each note is associated with one *user address* and one *app address*. Intuitively, a note belongs to the user address and is of a type given by the app address.
+Zero-knowledge proofs hide the data, but they aren't always enough to hide the validity predicates themselves (to provide functional privacy).
+The halo2 proving system is designed to handle this issue naturally, accumulating the proofs together.
 
-Currently, every [user](./users.md) in Taiga has two VPs: one that authorizes spending notes (`SendVP`), and one that authorizes receiving notes (`RecvVP`). [Apps](./app.md) in Taiga also have validity predicates, and spending or receiving a note of a particular app requires satisfying the `AppVP`.
-
-### Shielded VPs
-For each transaction, it must be proven that all of the VPs of parties and apps involved are satisfied. To preserve privacy, ZK proofs are used. The transaction is authorized only if all of the produced proofs are verified successfully.
-
-![img.png](img/vp_img.png)
+### VP interface
 
-### Transactions
+ZKP circuits have two types of inputs: public (visible to everyone) and public (only known to the prover).
+All validity predicates share the same *public input interface*, but can have different *private* inputs.
 
-Informally, transactions take a private subset of unspent notes from the Taiga note set, publicly reveal their nullifiers, and reveal a new set of note commitments to add to the Taiga note set. The Action/Execute circuit verifies consistency of this state transition, but does not check directly its validity. Instead, the validity predicate circuits must check the validity of the state transition. The following VPs are called:
+#### Public Inputs
 
-* The spending VP for every spent note
-* The app VP for every spent and created note
-* The receiving VP for every created note
+* $\{nf_i\}$ - the set of revealed nullifiers in the transaction
+* $\{cm_i\}$ - the set of new note commitments created in the transaction
+* $e$ - the current Taiga epoch (used for time-tracking)
```

Comment:
I don't think these are public inputs for VPs but for transactions

Reply:
I'm not sure it would make sense for transactions to have public or private inputs. It does make sense for the Action circuit, but we also have it for VPs, at least according to the current spec. The spec is outdated, of course, but it seemed like an intentional decision for both circuits to have nullifiers and commitments as a public input, and it feels like it wouldn't be affected much by what we've done since then, so I'm not sure how it is now. I'll clarify